lib/ai/chat.go

```diff
@@ -0,0 +1,250 @@
+				{Role: openai.ChatMessageRoleUser, Content: message},
+			},
+		},
+	)
+
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	return resp.Choices[0].Message.Content, nil
+}
+
+// StreamingMessage represents a message that is streamed from the assistant and will later be stored as a normal message in the conversation store.
+type StreamingMessage struct {
+	// Role describes the OpenAI role of the message, i.e it's sender.
+	Role string
+
+	// Idx is a semi-unique ID assigned when loading a conversation so that the UI can group partial messages together.
+	Idx int
+
+	// Chunks is a channel of message chunks that are streamed from the assistant.
+	Chunks <-chan string
+
+	// Error is a channel which may receive one error if the assistant encounters an error while streaming.
+	// Consumers should stop reading from all channels if they receive an error and abort.
+	Error <-chan error
+}
+
+// Complete completes the conversation with a message from the assistant based on the current context.
+// On success, it returns the message and the number of tokens used for the completion.
+func (chat *Chat) Complete(ctx context.Context) (any, int, error) {
+	var numTokens int
+
+	// if the chat is empty, return the initial response we predefine instead of querying GPT-4
+	if len(chat.messages) == 1 {
+		return &Message{
+			Role:    openai.ChatMessageRoleAssistant,
+			Content: initialAIResponse,
+			Idx:     len(chat.messages) - 1,
+		}, numTokens, nil
+	}
+
+	// if not, copy the current chat log to a new slice and append the suffix instruction
+	messages := make([]openai.ChatCompletionMessage, len(chat.messages)+1)
+	copy(messages, chat.messages)
+	messages[len(messages)-1] = openai.ChatCompletionMessage{
+		Role:    openai.ChatMessageRoleUser,
+		Content: promptExtractInstruction,
+	}
+
+	// create a streaming completion request, we do this to optimistically stream the response when
+	// we don't believe it's a payload
+	stream, err := chat.client.svc.CreateChatCompletionStream(
+		ctx,
+		openai.ChatCompletionRequest{
+			Model:     openai.GPT4,
+			Messages:  messages,
+			MaxTokens: maxResponseTokens,
+			Stream:    true,
+		},
+	)
+	if err != nil {
+		return nil, numTokens, trace.Wrap(err)
+	}
+
+	// fetch the first delta to check for a possible JSON payload
+	var response openai.ChatCompletionStreamResponse
+top:
+	response, err = stream.Recv()
+	if err != nil {
+		return nil, numTokens, trace.Wrap(err)
+	}
+	numTokens++
+
+	trimmed := strings.TrimSpace(response.Choices[0].Delta.Content)
+	if len(trimmed) == 0 {
+		goto top
+	}
+
+	// if it looks like a JSON payload, let's wait for the entire response and try to parse it
+	if strings.HasPrefix(trimmed, "{") {
+		payload := response.Choices[0].Delta.Content
+	outer:
+		for {
+			response, err := stream.Recv()
+			switch {
+			case errors.Is(err, io.EOF):
+				break outer
+			case err != nil:
+				return nil, numTokens, trace.Wrap(err)
+			}
```

Comment:
If you use just if/else if here instead of a switch, you don't need the label, and it's shorter:

```
if errors.Is(err, io.EOF) {
    break
} else if err != nil {
    return nil, numTokens, trace.Wrap(err)
}
```

Reply:
Changed